src/app/Fake.Sql.SqlServer/Sql.SqlServer.fs

```diff
@@ -0,0 +1,160 @@
+        Trace.logfn "Killing all processes from database %s on server %s." initialCatalog (getServerName serverInfo)
+        serverInfo.Server.KillAllProcesses initialCatalog
+        serverInfo
+
+    let detach serverInfo =
+        serverInfo
+        |> killAllProcesses
+        |> fun si ->
+                let initialCatalog = getInitialCatalog si
+                Trace.logfn "Detaching database %s on server %s." initialCatalog (getServerName si)
+                si.Server.DetachDatabase(initialCatalog, true)
+                si
+
+    let attach serverInfo (attachOptions:AttachOptions) files =
+        let sc = new Collections.Specialized.StringCollection()
+        files |> Seq.iter (fun file ->
+            sc.Add file |> ignore
+            File.checkExists file)
+
+        let initialCatalog = getInitialCatalog serverInfo
+
+        Trace.logfn "Attaching database %s on server %s." initialCatalog (getServerName serverInfo)
+        serverInfo.Server.AttachDatabase(initialCatalog, sc, attachOptions)
+        serverInfo
+
+    let createDb serverInfo =
+        Trace.logfn "Creating database %s on server %s." (getInitialCatalog serverInfo) (getServerName serverInfo)
+        (getDatabase serverInfo).Create()
+        serverInfo
+
+    let runScript serverInfo sqlFile =
+        Trace.logfn "Executing script %s." sqlFile
+        sqlFile
+        |> File.readAsString
+        |> (getDatabase serverInfo).ExecuteNonQuery
+
+    /// Closes the connection to the database server.
+    let disconnect serverInfo =
+        Trace.logfn "Disconnecting from server %s." (getServerName serverInfo)
+        if isNull serverInfo.Server then
+            failwith "Server is not configured."
+        if isNull serverInfo.Server.ConnectionContext then
+            failwith "Server.ConnectionContext is not configured."
+        serverInfo.Server.ConnectionContext.Disconnect()
+
+    let internal replaceDatabaseFilesF connectionString attachOptions copyF =
+        connectionString
+        |> getServerInfo
+        |> fun si -> if dbExistsOnServer si (getInitialCatalog si) then detach si else si
+        |> fun si -> copyF() |> attach si attachOptions
+        |> disconnect
+
+    /// Replaces the database files.
+    let replaceDatabaseFiles connectionString targetDir files attachOptions =
+        replaceDatabaseFilesF connectionString attachOptions
+            (fun _ ->
+                files
+                |> Seq.map (fun fileName ->
+                    let fi = new FileInfo(fileName)
+                    Shell.copyFile targetDir fileName
+                    targetDir @@ fi.Name))
+
+    let replaceDatabaseFilesWithCache connectionString targetDir cacheDir files attachOptions =
+        replaceDatabaseFilesF connectionString attachOptions
+            (fun _ -> Shell.copyCached targetDir cacheDir files)
+
+    let dropAndCreateDatabase connectionString =
+        connectionString
+        |> getServerInfo
+        |> dropDb
```

Comment:
I'm not sure if this are the signatures we want (ie returning `ServerInfo` instead of `unit`). This makes those methods weird to use on their own and you need `|> ignore` everywhere?

Not sure just my thinking around that

Reply:
I could not decide which way to go either, so I stuck with the signatures the old module had. 

I like returning the ServerInfo record from `attach`, `detach`, `killAllProcesses` because it indicates that something on the `serverInfo` record changed. The underlying calls to `serverInfo.Server.<Attach/Detach/KillAllProcesses>` read like they modify the `serverInfo.Server` object.

I see that those functions could return `unit` instead because `serverInfo` is already changed by `serverInfo.Server` mutating, and usually `unit` indicates some side effect happened.

Now that I write this all out, it sounds like returning `unit` is a clearer signature with the bonus of less `|> ignore`. I will get going on that and hopefully push it today.